pkg/tagger/remote/tagger.go

```diff
@@ -0,0 +1,270 @@
+	return t.store.subscribe(cardinality)
+}
+
+// Unsubscribe ends a subscription to entity events and closes its channel.
+func (t *Tagger) Unsubscribe(ch chan []types.EntityEvent) {
+	t.store.unsubscribe(ch)
+}
+
+func (t *Tagger) run() {
+	for {
+		select {
+		case <-t.health.C:
+		case <-t.ctx.Done():
+			return
+		default:
+		}
+
+		response, err := t.stream.Recv()
+		if err != nil {
+			log.Warnf("error received from remote tagger: %s", err)
+
+			// NOTE: when Recv() returns an error, the stream is
+			// aborted and needs to be re-established, so we do
+			// that here. also, startTaggerStream(0) will never
+			// return unless a stream can be established, or the
+			// tagger has been stopped, which means the error
+			// handling here is just a sanity check.
+			if err := t.startTaggerStream(0); err != nil {
+				log.Warnf("error received trying to start stream: %s", err)
+			}
+			continue
+		}
+
+		eventType, err := convertEventType(response.Type)
+		if err != nil {
+			log.Warnf("error processing event received from remote tagger: %s", err)
+			continue
+		}
+
+		err = t.store.processEvent(types.EntityEvent{
+			EventType: eventType,
+			Entity: types.Entity{
+				ID:                          convertEntityID(response.Entity.Id),
+				HighCardinalityTags:         response.Entity.HighCardinalityTags,
+				OrchestratorCardinalityTags: response.Entity.OrchestratorCardinalityTags,
+				LowCardinalityTags:          response.Entity.LowCardinalityTags,
+				StandardTags:                response.Entity.StandardTags,
+			},
+		})
+		if err != nil {
+			log.Warnf("error processing event received from remote tagger: %s", err)
+			continue
+		}
+	}
+}
+
+// startTaggerStream tries to establish a stream with the remote gRPC endpoint.
+// Since the entire remote tagger really depends on this working, it'll keep on
+// retrying with an exponential backoff until maxElapsed (or forever if
+// maxElapsed == 0) or the tagger is stopped.
+func (t *Tagger) startTaggerStream(maxElapsed time.Duration) error {
+	expBackoff := backoff.NewExponentialBackOff()
+	expBackoff.InitialInterval = 500 * time.Millisecond
+	expBackoff.MaxInterval = 5 * time.Minute
+	expBackoff.MaxElapsedTime = maxElapsed
+
+	return backoff.Retry(func() error {
+		select {
+		case <-t.ctx.Done():
+			return nil
+		default:
+		}
+
+		var err error
+		t.stream, err = t.client.TaggerStreamEntities(t.ctx, &pb.StreamTagsRequest{
+			Cardinality: pb.TagCardinality_HIGH,
+		})
+
+		if err != nil {
+			log.Infof("unable to establish stream, will possibly retry: %s", err)
+			return err
+		}
+
+		// when the stream is aborted, the store needs to be reset as
+		// its contents can no longer be relied to be up to date, and
+		// the new stream is responsible for re-adding all the still
+		// existing entities back. for the time in between, this can
+		// cause queries to the tagger to return nothing.
+		t.store.reset()
```

Comment:
Should we add a flag (internal) to flag remote readiness? In the spirit of `informers.WaitForCacheSync`, this would allow to log in `Tag()` requests received while tagger is not ready. It could be extremely useful in flares we get on support tickets.

Reply:
Good point. I think I'll have to change the gRPC protos to stream a list of events, instead of event by event, to be able to make this work. In that case, the first batch of events would mean the tagger is ready (otherwise, with event-by-event, we wouldn't know). That way I think we can also avoid the no tags case entirely, by keeping the old state of the store until we get a new batch. Changes incoming ;)